### PR TITLE
Provide connection between server and client

### PR DESCRIPTION
**Pull request by Artem-Semenov-dev**
This PR configures the ShareAware `gRPC` server and implements `ClientFacade` for easy interaction of the client with the server side of the application.

### CODE REVIEW

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            .onBehalfOf(user)
+            .subscribeToEvent(type)
+            .observe(observer)
+            .post()
+    }
+
+    /**
+     * Retrieves the entity with a provided type and ID.
+     */
+    public fun <E : EntityState> readEntity(type: Class<E>, id: Message): E {
```
Comment (Vladyslav-Kuksiuk) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:111]: What if we don't have an entity with this ID?

**Review by Vladyslav-Kuksiuk — COMMENTED** [review]
Mostly LGTM. @Artem-Semenov-dev please see my comment.

File: server/src/main/java/io/spine/examples/shareaware/server/ShareAwareServer.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        return Server
+                .atPort(DEFAULT_CLIENT_SERVICE_PORT)
+                .add(context)
+                .build();
+    }
+
+    private static void configureEnvironment() {
+        ServerEnvironment
+                .when(Production.class)
+                .use(InMemoryStorageFactory.newInstance())
```
Comment (armiol) [inline server/src/main/java/io/spine/examples/shareaware/server/ShareAwareServer.java:76]: We need to let users know about this somehow. I suggest to mention that in class-level documentation.

Otherwise it's not too clear why we lose all of our data upon each restart.

**Review by armiol — CHANGES_REQUESTED** [review]
@Artem-Semenov-dev please see my comments.

**Comment by armiol** [comment]
@Artem-Semenov-dev is this one still a draft?

**Comment by armiol** [comment]
@Artem-Semenov-dev as long as this one still is a draft, I am removing myself from the reviewers' duties.

**Comment by Artem-Semenov-dev** [comment]
> @Artem-Semenov-dev is this one still a draft?
Yes.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        client = Client
+            .usingChannel(channel)
+            .build()
+    }
+
+    /**
+     * Sends command to authenticate user in the system.
+     *
+     * A command `CreateWallet` is used for user creation because in the ShareAware
+     * there is no functional to register or authenticate the user, yet.
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:82]: "...because ShareAware server does not provide neither user registration nor authorisation features at the moment."

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    public fun authenticateUser() {
+        val userId = UUID.randomUUID().toUserId()
+        val walletId = userId.toWalletId()
+        val walletField = WalletCreated.Field.wallet()
+        val walletCreated: CompletableFuture<WalletCreated> = CompletableFuture()
+        this.subscribeToEvent(
+            WalletCreated::class.java,
+            EventFilter.eq(walletField, walletId)
+        )
+        {
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:93]: This seems to be some weird formatting.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+            walletCreated.complete(it)
+        }
+        command(createWallet(walletId))
+        if (!::user.isInitialized) {
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:97]: I am not sure why we use such a syntax here. Why not just call the method?

Same below.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+            .post()
+    }
+
+    /**
+     * Retrieves the entity with a provided type and ID.
+     *
+     * @param type type of the entity to be retrieved
+     * @param id entity ID by which the query result will be filtered
+     */
+    public fun <E : EntityState> readEntity(type: Class<E>, id: Message): E? {
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:153]: Let's document the returning value as well.

That's a part of a standard set of rules: public API should be documented. This is a part of public API. Please review this changeset for that.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        throw Exceptions.newIllegalStateException(
+            "User has not been authenticated in the system."
+        )
+    }
+
+    /**
+     * Returns the ID of the user's wallet if it exists.
+     *
+     * @throws IllegalStateException when the user's wallet does not exist
+     * because the authenticated user is not configured to the client.
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:182]: "...because there is authenticated user known to this client."?

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+     */
+    public fun wallet(): WalletId {
+        if (::walletId.isInitialized) {
+            return walletId
+        }
+        throw Exceptions.newIllegalStateException(
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:188]: Can we shorten this to just `newIllegalStateException`?

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+                    "the user has not been authenticated in the system."
+        )
+    }
+
+    /**
+     * Returns the `CreateWallet` command.
+     *
+     * @param wallet the ID of the wallet to create
+     */
+    private fun createWallet(wallet: WalletId): CreateWallet {
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:199]: Let's make this one an extension function to `CreateWallet`.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+        return CreateWallet
+            .newBuilder()
+            .setWallet(wallet)
+            .vBuild()
+    }
+
+    /**
+     * Converts the `UUID` to the `UserId`.
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:207]:
```
Creates a `UserId` taking a generated UUID value as a user identifier.
```

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+     */
+    private fun UUID.toUserId(): UserId {
+        return UserId
+            .newBuilder()
+            .setValue(this.toString())
+            .vBuild()
+    }
+
+    /**
+     * Converts the `UserId` to the `WalletId`.
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:217]: Please see my comment above on this matter.

We cannot convert one thing into another. They are not equivalent. We can just create something new on top of something that we posess.

File: server/src/main/java/io/spine/examples/shareaware/server/ShareAwareServer.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+import io.spine.server.delivery.Delivery;
+import io.spine.server.storage.memory.InMemoryStorageFactory;
+import io.spine.server.transport.memory.InMemoryTransportFactory;
+
+import java.io.IOException;
+
+import static io.spine.client.ConnectionConstants.DEFAULT_CLIENT_SERVICE_PORT;
+
+/**
+ * A "gRPC" server running a "Trading" bounded context.
```
Comment (armiol) [inline server/src/main/java/io/spine/examples/shareaware/server/ShareAwareServer.java:41]: Please don't quote gRPC. It's just a name of a library.

File: server/src/main/java/io/spine/examples/shareaware/server/ShareAwareServer.java [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * <p>ShareAware is an example application written on the `Spine` framework,
```
Comment (armiol) [inline server/src/main/java/io/spine/examples/shareaware/server/ShareAwareServer.java:43]: Same as above.

And I would rewrite it like this:
```
Server side of this application is currently running in in-memory storage mode. 
Therefore any changes made by users of this application will not be persisted
in-between the application launches.
```

**Review by armiol — CHANGES_REQUESTED** [review]
@Artem-Semenov-dev please see my comments.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt (hunk 5 above)
Comment (Artem-Semenov-dev) [reply client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:94]: I have defined this variable with `lateinit` property to avoid null instantiation. As Kotlin documentation says any call of this variable before its initialization will raise an exception. To find out if this variable is initialized or not we need to use property reference. 
<a href="https://kotlinlang.org/docs/whatsnew12.html#check-whether-a-lateinit-var-is-initialized:~:text=on%20v.1.8.21-,Check%20whether%20a%20lateinit%20var%20is%20initialized,-You%20can%20now">Documentation here</a>

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt (hunk 9 above)
Comment (Artem-Semenov-dev) [reply client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:199]: As far as I understand, it will be great to create an API like this `CreateWallet.withId(id)`. The problem is that the `CreateWallet` type like any other type from `Proto` files generated to Java and doesn't have a `Companion` object to add a static extension for. Of course, we can add an extension function for the `Builder` of the `CreateWallet` or add an extension for the `WalletId` type, but I think that is not what we want to do. 
There are some issues that I have found about static extensions: 
<a href="https://youtrack.jetbrains.com/issue/KT-11968">Issue on YouTrack</a>
<a href="https://github.com/Kotlin/KEEP/issues/348">Issue on GitHub</a>

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+            .setValue(this.toString())
+            .vBuild()
+    }
+
+    /**
+     * Creates a `WalletId` taking a `UserId` as a value for a wallet identifier.
+     *
+     * @return the ID of the user's wallet
+     */
+    private fun UserId.toWalletId(): WalletId {
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:221]: This should be the other way around. Wallet ID is created _from_ user ID.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt (hunk 9 above)
Comment (armiol) [reply client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:197]: Let's then have an extension to `Builder`.

@Vladyslav-Kuksiuk FYI.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+            .observe(observer)
+            .post()
+    }
+
+    /**
+     * Retrieves the entity with a provided type and ID.
+     *
+     * @param type type of the entity to be retrieved
+     * @param id entity ID by which the query result will be filtered
+     * @return the retrieved entity with provided ID and type if it exists otherwise null
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:149]:
```
..., `null` otherwise
```

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt (hunk 5 above)
Comment (armiol) [reply client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:94]: Ideally you don't need to check for that at all. Maybe, just once, somewhere during the application bootstrap.

The issue here is that `user` is never explicitly initialised in scope of this method prior to this check. Or it is, but I cannot see that.

How about we authenticate user right when the instance of `Client` is created? And then continue our journey knowing that `user` is always present?

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+
+    /**
+     * Sends command to authenticate user in the system.
+     *
+     * A command `CreateWallet` is used for user creation because ShareAware server
+     * does not provide neither user registration nor authorisation features at the moment.
+     */
+    public fun authenticateUser() {
+        val userId = UUID.randomUUID().toUserId()
+        val walletId = userId.toWalletId()
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:86]: I would split this method into parts. We create a wallet (which is a good candidate for extraction into something), and then we initialise ourselves upon the results arrival.

**Review by armiol — CHANGES_REQUESTED** [review]
@Artem-Semenov-dev please see my comments.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt (hunk 9 above)
Comment (armiol) [reply client/src/main/kotlin/io/spine/examples/shareaware/client/Client.kt:197]: @Artem-Semenov-dev @Vladyslav-Kuksiuk FYI.

I have tried this one myself, with the types I have in ProtoData. 

Here are the declarations I used:

```proto
message ProtobufSourceFile {
    option (entity).kind = VIEW;

    FilePath file_path = 1;

    // ...
}
```

```kotlin
private fun ProtobufSourceFile.takeMe(toChurch: String) : StringValue {
    return StringValue
        .newBuilder()
        .setValue(this.filePath.value + toChurch)
        .build()
}

public fun main() {
    val path = FilePath.newBuilder().setValue("/Earth/1973/Aerosmith/Dream")
    val protoFile = ProtobufSourceFile
        .newBuilder()
        .setFilePath(path)
        .build()

    val taken = protoFile.takeMe("On")
    println(taken.toJson())
}
```

And here is the output:
```
"/Earth/1973/Aerosmith/DreamOn"
```

**Review by armiol — APPROVED** [review]
@Artem-Semenov-dev Looks much better, I suggest to leave it like this for now.
